{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Query SRA metadata using Big Query\n",
    "\n",
    "Using https://pandas-gbq.readthedocs.io/. \n",
    "\n",
    "Before you begin, you must create a Google Cloud Platform project. Use the BigQuery sandbox to try the service for free.\n",
    "\n",
    "If you do not provide any credentials, this module attempts to load credentials from the environment. If no credentials are found, pandas-gbq prompts you to open a web browser, where you can grant it permissions to access your cloud resources. These credentials are only used locally."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas_gbq\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct query\n",
    "# Here we are using the bioproject PRJNA523380 for the CCLE cell lines\n",
    "# I only want to get the RNA-seq data \n",
    "query = \"\"\"\n",
    "SELECT * FROM `nih-sra-datastore.sra.metadata` \n",
    "WHERE bioproject = 'PRJNA523380' AND assay_type = 'RNA-Seq'\n",
    "LIMIT 5000\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Query the metadata table in BigQuery\n",
    "# Might take some time\n",
    "# Returns a dataframe\n",
    "df = pd.read_gbq(query, dialect=\"standard\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>center_name</th>\n",
       "      <th>consent</th>\n",
       "      <th>experiment</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>instrument</th>\n",
       "      <th>librarylayout</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>...</th>\n",
       "      <th>geo_loc_name_sam</th>\n",
       "      <th>ena_first_public_run</th>\n",
       "      <th>ena_last_update_run</th>\n",
       "      <th>sample_name_sam</th>\n",
       "      <th>datastore_filetype</th>\n",
       "      <th>datastore_provider</th>\n",
       "      <th>datastore_region</th>\n",
       "      <th>attributes</th>\n",
       "      <th>run_file_version</th>\n",
       "      <th>jattr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR8615408</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX5414345</td>\n",
       "      <td>AMO1_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[bam, run.zq, sra]</td>\n",
       "      <td>[gs, s3, ncbi]</td>\n",
       "      <td>[s3.us-east-1, gs.US, ncbi.public]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'female'}, {'k': 'asse...</td>\n",
       "      <td>1</td>\n",
       "      <td>{\"sex_calc\": \"female\", \"assemblyname\": \"GCA_00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8615898</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX5415152</td>\n",
       "      <td>NCIH1299_LUNG</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[bam, sra, run.zq]</td>\n",
       "      <td>[gs, s3, ncbi]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...</td>\n",
       "      <td>1</td>\n",
       "      <td>{\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR8616188</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX5414862</td>\n",
       "      <td>NCIH146_LUNG</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[sra, run.zq, bam]</td>\n",
       "      <td>[gs, s3, ncbi]</td>\n",
       "      <td>[ncbi.public, gs.US, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...</td>\n",
       "      <td>1</td>\n",
       "      <td>{\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR8615664</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX5414736</td>\n",
       "      <td>HUH1_LIVER</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[sra, bam, run.zq]</td>\n",
       "      <td>[gs, s3, ncbi]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...</td>\n",
       "      <td>1</td>\n",
       "      <td>{\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR8615310</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX5414443</td>\n",
       "      <td>SNU398_LIVER</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[run.zq, bam, sra]</td>\n",
       "      <td>[gs, s3, ncbi]</td>\n",
       "      <td>[gs.US, s3.us-east-1, ncbi.public]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...</td>\n",
       "      <td>1</td>\n",
       "      <td>{\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1014</th>\n",
       "      <td>SRR8615714</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX5414686</td>\n",
       "      <td>KALS1_CENTRAL_NERVOUS_SYSTEM</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[run.zq, sra, bam]</td>\n",
       "      <td>[ncbi, s3, gs]</td>\n",
       "      <td>[s3.us-east-1, gs.US, ncbi.public]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'female'}, {'k': 'asse...</td>\n",
       "      <td>1</td>\n",
       "      <td>{\"sex_calc\": \"female\", \"assemblyname\": \"GCA_00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1015</th>\n",
       "      <td>SRR8615590</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX5414810</td>\n",
       "      <td>RERFLCKJ_LUNG</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[run.zq, sra, bam]</td>\n",
       "      <td>[ncbi, s3, gs]</td>\n",
       "      <td>[gs.US, s3.us-east-1, ncbi.public]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...</td>\n",
       "      <td>1</td>\n",
       "      <td>{\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1016</th>\n",
       "      <td>SRR8615630</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX5414770</td>\n",
       "      <td>HEL_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[bam, run.zq, sra]</td>\n",
       "      <td>[ncbi, s3, gs]</td>\n",
       "      <td>[ncbi.public, gs.US, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...</td>\n",
       "      <td>1</td>\n",
       "      <td>{\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1017</th>\n",
       "      <td>SRR8615730</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX5414670</td>\n",
       "      <td>TTC549_SOFT_TISSUE</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[sra, bam, run.zq]</td>\n",
       "      <td>[ncbi, s3, gs]</td>\n",
       "      <td>[gs.US, s3.us-east-1, ncbi.public]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'NA'}, {'k': 'assembly...</td>\n",
       "      <td>1</td>\n",
       "      <td>{\"sex_calc\": \"NA\", \"assemblyname\": \"GCA_000001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1018</th>\n",
       "      <td>SRR8615885</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX5415165</td>\n",
       "      <td>HS934T_FIBROBLAST</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[bam, sra, run.zq]</td>\n",
       "      <td>[ncbi, s3, gs]</td>\n",
       "      <td>[gs.US, s3.us-east-1, ncbi.public]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'female'}, {'k': 'asse...</td>\n",
       "      <td>1</td>\n",
       "      <td>{\"sex_calc\": \"female\", \"assemblyname\": \"GCA_00...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1019 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             acc assay_type      center_name consent  experiment  \\\n",
       "0     SRR8615408    RNA-Seq  BROAD INSTITUTE  public  SRX5414345   \n",
       "1     SRR8615898    RNA-Seq  BROAD INSTITUTE  public  SRX5415152   \n",
       "2     SRR8616188    RNA-Seq  BROAD INSTITUTE  public  SRX5414862   \n",
       "3     SRR8615664    RNA-Seq  BROAD INSTITUTE  public  SRX5414736   \n",
       "4     SRR8615310    RNA-Seq  BROAD INSTITUTE  public  SRX5414443   \n",
       "...          ...        ...              ...     ...         ...   \n",
       "1014  SRR8615714    RNA-Seq  BROAD INSTITUTE  public  SRX5414686   \n",
       "1015  SRR8615590    RNA-Seq  BROAD INSTITUTE  public  SRX5414810   \n",
       "1016  SRR8615630    RNA-Seq  BROAD INSTITUTE  public  SRX5414770   \n",
       "1017  SRR8615730    RNA-Seq  BROAD INSTITUTE  public  SRX5414670   \n",
       "1018  SRR8615885    RNA-Seq  BROAD INSTITUTE  public  SRX5415165   \n",
       "\n",
       "                                  sample_name           instrument  \\\n",
       "0     AMO1_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE  Illumina HiSeq 2500   \n",
       "1                               NCIH1299_LUNG  Illumina HiSeq 2500   \n",
       "2                                NCIH146_LUNG  Illumina HiSeq 2500   \n",
       "3                                  HUH1_LIVER  Illumina HiSeq 2500   \n",
       "4                                SNU398_LIVER  Illumina HiSeq 2500   \n",
       "...                                       ...                  ...   \n",
       "1014             KALS1_CENTRAL_NERVOUS_SYSTEM  Illumina HiSeq 2500   \n",
       "1015                            RERFLCKJ_LUNG  Illumina HiSeq 2500   \n",
       "1016   HEL_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE  Illumina HiSeq 2500   \n",
       "1017                       TTC549_SOFT_TISSUE  Illumina HiSeq 2500   \n",
       "1018                        HS934T_FIBROBLAST  Illumina HiSeq 2500   \n",
       "\n",
       "     librarylayout libraryselection   librarysource  ... geo_loc_name_sam  \\\n",
       "0           PAIRED             cDNA  TRANSCRIPTOMIC  ...               []   \n",
       "1           PAIRED             cDNA  TRANSCRIPTOMIC  ...               []   \n",
       "2           PAIRED             cDNA  TRANSCRIPTOMIC  ...               []   \n",
       "3           PAIRED             cDNA  TRANSCRIPTOMIC  ...               []   \n",
       "4           PAIRED             cDNA  TRANSCRIPTOMIC  ...               []   \n",
       "...            ...              ...             ...  ...              ...   \n",
       "1014        PAIRED             cDNA  TRANSCRIPTOMIC  ...               []   \n",
       "1015        PAIRED             cDNA  TRANSCRIPTOMIC  ...               []   \n",
       "1016        PAIRED             cDNA  TRANSCRIPTOMIC  ...               []   \n",
       "1017        PAIRED             cDNA  TRANSCRIPTOMIC  ...               []   \n",
       "1018        PAIRED             cDNA  TRANSCRIPTOMIC  ...               []   \n",
       "\n",
       "     ena_first_public_run ena_last_update_run sample_name_sam  \\\n",
       "0                      []                  []              []   \n",
       "1                      []                  []              []   \n",
       "2                      []                  []              []   \n",
       "3                      []                  []              []   \n",
       "4                      []                  []              []   \n",
       "...                   ...                 ...             ...   \n",
       "1014                   []                  []              []   \n",
       "1015                   []                  []              []   \n",
       "1016                   []                  []              []   \n",
       "1017                   []                  []              []   \n",
       "1018                   []                  []              []   \n",
       "\n",
       "      datastore_filetype datastore_provider  \\\n",
       "0     [bam, run.zq, sra]     [gs, s3, ncbi]   \n",
       "1     [bam, sra, run.zq]     [gs, s3, ncbi]   \n",
       "2     [sra, run.zq, bam]     [gs, s3, ncbi]   \n",
       "3     [sra, bam, run.zq]     [gs, s3, ncbi]   \n",
       "4     [run.zq, bam, sra]     [gs, s3, ncbi]   \n",
       "...                  ...                ...   \n",
       "1014  [run.zq, sra, bam]     [ncbi, s3, gs]   \n",
       "1015  [run.zq, sra, bam]     [ncbi, s3, gs]   \n",
       "1016  [bam, run.zq, sra]     [ncbi, s3, gs]   \n",
       "1017  [sra, bam, run.zq]     [ncbi, s3, gs]   \n",
       "1018  [bam, sra, run.zq]     [ncbi, s3, gs]   \n",
       "\n",
       "                        datastore_region  \\\n",
       "0     [s3.us-east-1, gs.US, ncbi.public]   \n",
       "1     [gs.US, ncbi.public, s3.us-east-1]   \n",
       "2     [ncbi.public, gs.US, s3.us-east-1]   \n",
       "3     [gs.US, ncbi.public, s3.us-east-1]   \n",
       "4     [gs.US, s3.us-east-1, ncbi.public]   \n",
       "...                                  ...   \n",
       "1014  [s3.us-east-1, gs.US, ncbi.public]   \n",
       "1015  [gs.US, s3.us-east-1, ncbi.public]   \n",
       "1016  [ncbi.public, gs.US, s3.us-east-1]   \n",
       "1017  [gs.US, s3.us-east-1, ncbi.public]   \n",
       "1018  [gs.US, s3.us-east-1, ncbi.public]   \n",
       "\n",
       "                                             attributes run_file_version  \\\n",
       "0     [{'k': 'sex_calc', 'v': 'female'}, {'k': 'asse...                1   \n",
       "1     [{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...                1   \n",
       "2     [{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...                1   \n",
       "3     [{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...                1   \n",
       "4     [{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...                1   \n",
       "...                                                 ...              ...   \n",
       "1014  [{'k': 'sex_calc', 'v': 'female'}, {'k': 'asse...                1   \n",
       "1015  [{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...                1   \n",
       "1016  [{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...                1   \n",
       "1017  [{'k': 'sex_calc', 'v': 'NA'}, {'k': 'assembly...                1   \n",
       "1018  [{'k': 'sex_calc', 'v': 'female'}, {'k': 'asse...                1   \n",
       "\n",
       "                                                  jattr  \n",
       "0     {\"sex_calc\": \"female\", \"assemblyname\": \"GCA_00...  \n",
       "1     {\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...  \n",
       "2     {\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...  \n",
       "3     {\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...  \n",
       "4     {\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...  \n",
       "...                                                 ...  \n",
       "1014  {\"sex_calc\": \"female\", \"assemblyname\": \"GCA_00...  \n",
       "1015  {\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...  \n",
       "1016  {\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...  \n",
       "1017  {\"sex_calc\": \"NA\", \"assemblyname\": \"GCA_000001...  \n",
       "1018  {\"sex_calc\": \"female\", \"assemblyname\": \"GCA_00...  \n",
       "\n",
       "[1019 rows x 37 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Goal here is to curate metadata for the results of the query\n",
    "I specifically want a dataframe with the desired metadata as well as a dictionary with the same information.\n",
    "\n",
    "Hoping to be able to use the dictionary to add the metadata to the objects in the the GCS bucket "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  TODO:: refactor these functions to be more generic\n",
    "# Also improve efficiency by not iterating over the entire dataframe\n",
    "# Maybe use a dictionary comprehension instead\n",
    "\n",
    "def convert_array_to_dict(arr):\n",
    "    result_dict = {}\n",
    "    desired_keys = ['bases', 'bytes', 'run_file_create_date', 'disease_sam', 'disease_stage_sam_s_dpl172', 'tissue_sam']\n",
    "    for item in arr:\n",
    "        if isinstance(item, dict):\n",
    "            key = item.get('k')\n",
    "            value = item.get('v')\n",
    "            if key in desired_keys:\n",
    "                if key == 'bytes':\n",
    "                    result_dict['size_in_bytes'] = int(value)\n",
    "                    result_dict['size_in_GB'] = round(float(value) / 1000000000, 2)\n",
    "                elif key == 'run_file_create_date':\n",
    "                    result_dict[key] = str(value)\n",
    "                else:\n",
    "                    if key in result_dict:\n",
    "                        if isinstance(result_dict[key], list):\n",
    "                            result_dict[key].append(value)\n",
    "                        else:\n",
    "                            result_dict[key] = [result_dict[key], value]\n",
    "                    else:\n",
    "                        result_dict[key] = value\n",
    "    return result_dict\n",
    "\n",
    "def convert_row_to_dict(row):\n",
    "    result_dict = {}\n",
    "    for column in row.index:\n",
    "        if column == 'attributes':\n",
    "            result_dict.update(convert_array_to_dict(row[column]))\n",
    "        elif column == 'releasedate':\n",
    "            # convert type Timestamp to string\n",
    "            result_dict[column] = str(row[column])\n",
    "        elif column == 'run_file_create_date':\n",
    "            result_dict[column] = str(row[column])\n",
    "        else:\n",
    "            result_dict[column] = row[column]\n",
    "    return result_dict\n",
    "\n",
    "\n",
    "def convert_dataframe_to_dict(df):\n",
    "    result_dict = {}\n",
    "    for i in range(len(df)):\n",
    "        result_dict[df['acc'][i]] = convert_row_to_dict(df.iloc[i])\n",
    "    return result_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "acc has 1019 unique values. \n",
      "Examples: ['SRR8615408' 'SRR8615898' 'SRR8616188' 'SRR8615664' 'SRR8615310']\n",
      "\n",
      "sample_name has 1019 unique values. \n",
      "Examples: ['AMO1_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE' 'NCIH1299_LUNG' 'NCIH146_LUNG'\n",
      " 'HUH1_LIVER' 'SNU398_LIVER']\n",
      "\n",
      "sample_acc has 1019 unique values. \n",
      "Examples: ['SRS4395396' 'SRS4396049' 'SRS4395814' 'SRS4395725' 'SRS4395475']\n",
      "\n",
      "experiment has 1019 unique values. \n",
      "Examples: ['SRX5414345' 'SRX5415152' 'SRX5414862' 'SRX5414736' 'SRX5414443']\n",
      "\n",
      "library_name has 1019 unique values. \n",
      "Examples: ['RNASeq-AMO1_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE' 'RNASeq-NCIH1299_LUNG'\n",
      " 'RNASeq-NCIH146_LUNG' 'RNASeq-HUH1_LIVER' 'RNASeq-SNU398_LIVER']\n",
      "\n",
      "sra_study has 1 unique values. \n",
      "Examples: ['SRP186687']\n",
      "\n",
      "center_name has 1 unique values. \n",
      "Examples: ['BROAD INSTITUTE']\n",
      "\n",
      "platform has 1 unique values. \n",
      "Examples: ['ILLUMINA']\n",
      "\n",
      "assay_type has 1 unique values. \n",
      "Examples: ['RNA-Seq']\n",
      "\n",
      "librarysource has 1 unique values. \n",
      "Examples: ['TRANSCRIPTOMIC']\n",
      "\n",
      "organism has 1 unique values. \n",
      "Examples: ['Homo sapiens']\n",
      "\n",
      "releasedate has 1 unique values. \n",
      "Examples: <DatetimeArray>\n",
      "['2019-03-27 00:00:00+00:00']\n",
      "Length: 1, dtype: datetime64[us, UTC]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# choose columns\n",
    "columns = ['acc', 'sample_name', 'sample_acc', 'experiment',  'library_name', 'sra_study', 'center_name', \n",
    "'platform', 'assay_type', 'librarysource', 'organism', 'releasedate']\n",
    "\n",
    "# subset the dataframe to only include the columns we want\n",
    "df_ = df[columns + ['attributes']].copy()\n",
    "\n",
    "# # for each column in columns, print out the number of unique values, and then the first 5 unique values\n",
    "for col in columns:\n",
    "    # print(f\"For {col}, there are: {df[col].nunique()} unique values. \\nExamples: {df[col].unique()[0:5]}\\n\")\n",
    "    print(f\"{col} has {df[col].nunique()} unique values. \\nExamples: {df[col].unique()[0:5]}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset df to df_ using columns but also include the 'attributes' column\n",
    "dict_metadata = convert_dataframe_to_dict(df_)\n",
    "dict_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert dict to pandas dataframe\n",
    "df_metadata = pd.DataFrame.from_dict(dict_metadata, orient='index')\n",
    "# rename the 'acc' column to 'run_accession'\n",
    "df_metadata.rename(columns={'acc': 'run_accession'}, inplace=True)\n",
    "df_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # save the dataframe to a csv file \n",
    "# df_metadata.to_csv('../../metadata/sra_metadata.csv', index=False)\n",
    "\n",
    "# # save the dictionary to a json file\n",
    "# import json\n",
    "# with open('../../metadata/sra_metadata.json', 'w') as fp:\n",
    "#     json.dump(dict_metadata, fp)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "getSRA",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
