{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "from json import loads, dumps\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create sql query\n",
    "- TODO:: find out how to get the schema of a BigQuery table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "sql = \"\"\"\n",
    "    SELECT *\n",
    "    FROM `nih-sra-datastore.sra.metadata`\n",
    "    WHERE bioproject=\"PRJNA523380\" AND center_name = \"BROAD INSTITUTE\"\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pandas.read_gbq(sql, dialect=\"standard\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>center_name</th>\n",
       "      <th>consent</th>\n",
       "      <th>experiment</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>instrument</th>\n",
       "      <th>librarylayout</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>...</th>\n",
       "      <th>geo_loc_name_sam</th>\n",
       "      <th>ena_first_public_run</th>\n",
       "      <th>ena_last_update_run</th>\n",
       "      <th>sample_name_sam</th>\n",
       "      <th>datastore_filetype</th>\n",
       "      <th>datastore_provider</th>\n",
       "      <th>datastore_region</th>\n",
       "      <th>attributes</th>\n",
       "      <th>run_file_version</th>\n",
       "      <th>jattr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR8652105</td>\n",
       "      <td>WGS</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX5449795</td>\n",
       "      <td>MCF7_BREAST</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[bam, sra, run.zq]</td>\n",
       "      <td>[ncbi, s3, gs]</td>\n",
       "      <td>[ncbi.public, gs.US, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'female'}, {'k': 'asse...</td>\n",
       "      <td>1</td>\n",
       "      <td>{\"sex_calc\": \"female\", \"assemblyname\": \"GCA_00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8639219</td>\n",
       "      <td>WGS</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX5437514</td>\n",
       "      <td>DMS53_LUNG</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[bam, sra, run.zq]</td>\n",
       "      <td>[s3, ncbi, gs]</td>\n",
       "      <td>[ncbi.public, s3.us-east-1, gs.US]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...</td>\n",
       "      <td>1</td>\n",
       "      <td>{\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR8652062</td>\n",
       "      <td>WGS</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX5449838</td>\n",
       "      <td>IM95_STOMACH</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[run.zq, bam, sra]</td>\n",
       "      <td>[gs, s3, ncbi]</td>\n",
       "      <td>[s3.us-east-1, gs.US, ncbi.public]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...</td>\n",
       "      <td>1</td>\n",
       "      <td>{\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR8652060</td>\n",
       "      <td>WGS</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX5449840</td>\n",
       "      <td>KATOIII_STOMACH</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[run.zq, bam, sra]</td>\n",
       "      <td>[gs, s3, ncbi]</td>\n",
       "      <td>[s3.us-east-1, ncbi.public, gs.US]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...</td>\n",
       "      <td>1</td>\n",
       "      <td>{\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR8652054</td>\n",
       "      <td>WGS</td>\n",
       "      <td>BROAD INSTITUTE</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX5449846</td>\n",
       "      <td>NCIH2066_LUNG</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[sra, run.zq, bam]</td>\n",
       "      <td>[s3, ncbi, gs]</td>\n",
       "      <td>[ncbi.public, gs.US, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'female'}, {'k': 'asse...</td>\n",
       "      <td>1</td>\n",
       "      <td>{\"sex_calc\": \"female\", \"assemblyname\": \"GCA_00...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          acc assay_type      center_name consent  experiment  \\\n",
       "0  SRR8652105        WGS  BROAD INSTITUTE  public  SRX5449795   \n",
       "1  SRR8639219        WGS  BROAD INSTITUTE  public  SRX5437514   \n",
       "2  SRR8652062        WGS  BROAD INSTITUTE  public  SRX5449838   \n",
       "3  SRR8652060        WGS  BROAD INSTITUTE  public  SRX5449840   \n",
       "4  SRR8652054        WGS  BROAD INSTITUTE  public  SRX5449846   \n",
       "\n",
       "       sample_name           instrument librarylayout    libraryselection  \\\n",
       "0      MCF7_BREAST  Illumina HiSeq 2000        PAIRED  size fractionation   \n",
       "1       DMS53_LUNG  Illumina HiSeq 2000        PAIRED  size fractionation   \n",
       "2     IM95_STOMACH  Illumina HiSeq 2000        PAIRED  size fractionation   \n",
       "3  KATOIII_STOMACH  Illumina HiSeq 2000        PAIRED  size fractionation   \n",
       "4    NCIH2066_LUNG  Illumina HiSeq 2000        PAIRED  size fractionation   \n",
       "\n",
       "  librarysource  ... geo_loc_name_sam ena_first_public_run  \\\n",
       "0       GENOMIC  ...               []                   []   \n",
       "1       GENOMIC  ...               []                   []   \n",
       "2       GENOMIC  ...               []                   []   \n",
       "3       GENOMIC  ...               []                   []   \n",
       "4       GENOMIC  ...               []                   []   \n",
       "\n",
       "  ena_last_update_run sample_name_sam  datastore_filetype datastore_provider  \\\n",
       "0                  []              []  [bam, sra, run.zq]     [ncbi, s3, gs]   \n",
       "1                  []              []  [bam, sra, run.zq]     [s3, ncbi, gs]   \n",
       "2                  []              []  [run.zq, bam, sra]     [gs, s3, ncbi]   \n",
       "3                  []              []  [run.zq, bam, sra]     [gs, s3, ncbi]   \n",
       "4                  []              []  [sra, run.zq, bam]     [s3, ncbi, gs]   \n",
       "\n",
       "                     datastore_region  \\\n",
       "0  [ncbi.public, gs.US, s3.us-east-1]   \n",
       "1  [ncbi.public, s3.us-east-1, gs.US]   \n",
       "2  [s3.us-east-1, gs.US, ncbi.public]   \n",
       "3  [s3.us-east-1, ncbi.public, gs.US]   \n",
       "4  [ncbi.public, gs.US, s3.us-east-1]   \n",
       "\n",
       "                                          attributes run_file_version  \\\n",
       "0  [{'k': 'sex_calc', 'v': 'female'}, {'k': 'asse...                1   \n",
       "1  [{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...                1   \n",
       "2  [{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...                1   \n",
       "3  [{'k': 'sex_calc', 'v': 'male'}, {'k': 'assemb...                1   \n",
       "4  [{'k': 'sex_calc', 'v': 'female'}, {'k': 'asse...                1   \n",
       "\n",
       "                                               jattr  \n",
       "0  {\"sex_calc\": \"female\", \"assemblyname\": \"GCA_00...  \n",
       "1  {\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...  \n",
       "2  {\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...  \n",
       "3  {\"sex_calc\": \"male\", \"assemblyname\": \"GCA_0000...  \n",
       "4  {\"sex_calc\": \"female\", \"assemblyname\": \"GCA_00...  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# View dataframe\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analayze df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['acc', 'assay_type', 'center_name', 'consent', 'experiment',\n",
      "       'sample_name', 'instrument', 'librarylayout', 'libraryselection',\n",
      "       'librarysource', 'platform', 'sample_acc', 'biosample', 'organism',\n",
      "       'sra_study', 'releasedate', 'bioproject', 'mbytes', 'loaddate',\n",
      "       'avgspotlen', 'mbases', 'insertsize', 'library_name',\n",
      "       'biosamplemodel_sam', 'collection_date_sam',\n",
      "       'geo_loc_name_country_calc', 'geo_loc_name_country_continent_calc',\n",
      "       'geo_loc_name_sam', 'ena_first_public_run', 'ena_last_update_run',\n",
      "       'sample_name_sam', 'datastore_filetype', 'datastore_provider',\n",
      "       'datastore_region', 'attributes', 'run_file_version', 'jattr'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique entries for column 'organism':\n",
      "       organism  count\n",
      "0  Homo sapiens   4550\n",
      "\n",
      "\n",
      "Unique entries for column 'instrument':\n",
      "            instrument  count\n",
      "0  Illumina HiSeq 2000   3341\n",
      "1  Illumina HiSeq 2500   1019\n",
      "2       Illumina MiSeq    190\n",
      "\n",
      "\n",
      "Unique entries for column 'assay_type':\n",
      "         assay_type  count\n",
      "0          AMPLICON    972\n",
      "1     Bisulfite-Seq    928\n",
      "2           RNA-Seq   1019\n",
      "3  Targeted-Capture    976\n",
      "4               WGS    329\n",
      "5               WXS    326\n",
      "\n",
      "\n",
      "Unique entries for column 'librarylayout':\n",
      "  librarylayout  count\n",
      "0        PAIRED   3622\n",
      "1        SINGLE    928\n",
      "\n",
      "\n",
      "Unique entries for column 'libraryselection':\n",
      "         libraryselection  count\n",
      "0        Hybrid Selection   1302\n",
      "1                     PCR    972\n",
      "2  Reduced Representation    928\n",
      "3                    cDNA   1019\n",
      "4      size fractionation    329\n",
      "\n",
      "\n",
      "Unique entries for column 'librarysource':\n",
      "    librarysource  count\n",
      "0         GENOMIC   3531\n",
      "1  TRANSCRIPTOMIC   1019\n",
      "\n",
      "\n",
      "Unique entries for column 'releasedate':\n",
      "                releasedate  count\n",
      "0 2019-03-27 00:00:00+00:00   4550\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Assuming df is your original DataFrame containing the data\n",
    "\n",
    "# List of column names for which you want to create unique DataFrames\n",
    "columns_to_extract = [\"organism\", \"instrument\", \"assay_type\", \"librarylayout\", \"libraryselection\", \"librarysource\", \"releasedate\"]\n",
    "\n",
    "# Create a dictionary of DataFrames with counts of unique values for each column\n",
    "unique_dataframes = {\n",
    "    column: df.groupby(column).size().reset_index(name='count') for column in columns_to_extract\n",
    "}\n",
    "\n",
    "# Printing each DataFrame\n",
    "for column, unique_df in unique_dataframes.items():\n",
    "    print(f\"Unique entries for column '{column}':\")\n",
    "    print(unique_df)\n",
    "    print(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'organism':        organism  count\n",
       " 0  Homo sapiens   4550,\n",
       " 'instrument':             instrument  count\n",
       " 0  Illumina HiSeq 2000   3341\n",
       " 1  Illumina HiSeq 2500   1019\n",
       " 2       Illumina MiSeq    190,\n",
       " 'assay_type':          assay_type  count\n",
       " 0          AMPLICON    972\n",
       " 1     Bisulfite-Seq    928\n",
       " 2           RNA-Seq   1019\n",
       " 3  Targeted-Capture    976\n",
       " 4               WGS    329\n",
       " 5               WXS    326,\n",
       " 'librarylayout':   librarylayout  count\n",
       " 0        PAIRED   3622\n",
       " 1        SINGLE    928,\n",
       " 'libraryselection':          libraryselection  count\n",
       " 0        Hybrid Selection   1302\n",
       " 1                     PCR    972\n",
       " 2  Reduced Representation    928\n",
       " 3                    cDNA   1019\n",
       " 4      size fractionation    329,\n",
       " 'librarysource':     librarysource  count\n",
       " 0         GENOMIC   3531\n",
       " 1  TRANSCRIPTOMIC   1019,\n",
       " 'releasedate':                 releasedate  count\n",
       " 0 2019-03-27 00:00:00+00:00   4550}"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_dataframes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SAVE JSON to ./data/ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "result=df.to_json(orient=\"records\")\n",
    "parsed = loads(result)\n",
    "x=dumps(parsed, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "def write_json_to_file(file_path, data):\n",
    "    # Convert the file path to a Path object for better path manipulation\n",
    "    path = Path(file_path)\n",
    "    \n",
    "    # Create the parent directory if it doesn't exist\n",
    "    path.parent.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "    # Write the data to the file\n",
    "    with open(path, \"w\") as outfile:\n",
    "        outfile.write(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath=\"./data/BQ_CCLE_metadata.json\"\n",
    "write_json_to_file(filepath, x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'acc': 'SRR8639219',\n",
       " 'assay_type': 'WGS',\n",
       " 'center_name': 'BROAD INSTITUTE',\n",
       " 'consent': 'public',\n",
       " 'experiment': 'SRX5437514',\n",
       " 'sample_name': 'DMS53_LUNG',\n",
       " 'instrument': 'Illumina HiSeq 2000',\n",
       " 'librarylayout': 'PAIRED',\n",
       " 'libraryselection': 'size fractionation',\n",
       " 'librarysource': 'GENOMIC',\n",
       " 'platform': 'ILLUMINA',\n",
       " 'sample_acc': 'SRS4395966',\n",
       " 'biosample': 'SAMN10987596',\n",
       " 'organism': 'Homo sapiens',\n",
       " 'sra_study': 'SRP186687',\n",
       " 'releasedate': 1553644800000,\n",
       " 'bioproject': 'PRJNA523380',\n",
       " 'mbytes': 69535,\n",
       " 'loaddate': None,\n",
       " 'avgspotlen': 202,\n",
       " 'mbases': 135250,\n",
       " 'insertsize': None,\n",
       " 'library_name': 'WGS-DMS53_LUNG',\n",
       " 'biosamplemodel_sam': ['Human'],\n",
       " 'collection_date_sam': None,\n",
       " 'geo_loc_name_country_calc': None,\n",
       " 'geo_loc_name_country_continent_calc': None,\n",
       " 'geo_loc_name_sam': [],\n",
       " 'ena_first_public_run': [],\n",
       " 'ena_last_update_run': [],\n",
       " 'sample_name_sam': [],\n",
       " 'datastore_filetype': ['bam', 'sra', 'run.zq'],\n",
       " 'datastore_provider': ['s3', 'ncbi', 'gs'],\n",
       " 'datastore_region': ['ncbi.public', 's3.us-east-1', 'gs.US'],\n",
       " 'attributes': [{'k': 'sex_calc', 'v': 'male'},\n",
       "  {'k': 'assemblyname', 'v': 'GCA_000001405.13'},\n",
       "  {'k': 'bases', 'v': '135250787508'},\n",
       "  {'k': 'bytes', 'v': '72913109560'},\n",
       "  {'k': 'run_file_create_date', 'v': '2019-02-27T00:35:00.000Z'},\n",
       "  {'k': 'age_sam', 'v': '54'},\n",
       "  {'k': 'biomaterial_provider_sam', 'v': 'ATCC:DMS 53'},\n",
       "  {'k': 'cell_line_sam_ss_dpl110', 'v': 'DMS 53'},\n",
       "  {'k': 'disease_sam', 'v': 'carcinoma (small_cell_carcinoma)'},\n",
       "  {'k': 'disease_stage_sam_s_dpl172', 'v': 'primary'},\n",
       "  {'k': 'ethnicity_sam', 'v': 'Caucasian'},\n",
       "  {'k': 'isolate_sam', 'v': 'cell line'},\n",
       "  {'k': 'sample_type_sam', 'v': 'cell culture'},\n",
       "  {'k': 'tissue_sam', 'v': 'lung'},\n",
       "  {'k': 'primary_search', 'v': '10987596'},\n",
       "  {'k': 'primary_search', 'v': '523380'},\n",
       "  {'k': 'primary_search', 'v': 'DMS53_LUNG'},\n",
       "  {'k': 'primary_search', 'v': 'DMS_53.bam'},\n",
       "  {'k': 'primary_search', 'v': 'PRJNA152953'},\n",
       "  {'k': 'primary_search', 'v': 'PRJNA523380'},\n",
       "  {'k': 'primary_search', 'v': 'SAMN10987596'},\n",
       "  {'k': 'primary_search', 'v': 'SRP186687'},\n",
       "  {'k': 'primary_search', 'v': 'SRR8639219'},\n",
       "  {'k': 'primary_search', 'v': 'SRS4395966'},\n",
       "  {'k': 'primary_search', 'v': 'SRX5437514'},\n",
       "  {'k': 'primary_search', 'v': 'WGS-DMS53_LUNG'}],\n",
       " 'run_file_version': 1,\n",
       " 'jattr': '{\"sex_calc\": \"male\", \"assemblyname\": \"GCA_000001405.13\", \"bases\": 135250787508, \"bytes\": 72913109560, \"run_file_create_date\": \"2019-02-27T00:35:00.000Z\", \"age_sam\": [\"54\"], \"biomaterial_provider_sam\": [\"ATCC:DMS 53\"], \"cell_line_sam_ss_dpl110\": [\"DMS 53\"], \"disease_sam\": [\"carcinoma (small_cell_carcinoma)\"], \"disease_stage_sam_s_dpl172\": \"primary\", \"ethnicity_sam\": [\"Caucasian\"], \"isolate_sam\": [\"cell line\"], \"sample_type_sam\": [\"cell culture\"], \"tissue_sam\": [\"lung\"], \"primary_search\": \"10987596\"}'}"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loads(x)[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>mbytes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>695</th>\n",
       "      <td>SRR8615894</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>TUHR4TKB_KIDNEY</td>\n",
       "      <td>3904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048</th>\n",
       "      <td>SRR8618300</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>VMCUB1_URINARY_TRACT</td>\n",
       "      <td>3561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1128</th>\n",
       "      <td>SRR8615581</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>MDAMB361_BREAST</td>\n",
       "      <td>2995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1673</th>\n",
       "      <td>SRR8615564</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SW1463_LARGE_INTESTINE</td>\n",
       "      <td>3960</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             acc   bioproject             sample_name  mbytes\n",
       "695   SRR8615894  PRJNA523380         TUHR4TKB_KIDNEY    3904\n",
       "1048  SRR8618300  PRJNA523380    VMCUB1_URINARY_TRACT    3561\n",
       "1128  SRR8615581  PRJNA523380         MDAMB361_BREAST    2995\n",
       "1673  SRR8615564  PRJNA523380  SW1463_LARGE_INTESTINE    3960"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df[\"mbytes\"] >= 1000) & (df[\"assay_type\"]==\"RNA-Seq\") &(df[\"mbytes\"] <= 4000)][[\"acc\",\"bioproject\", \"sample_name\", \"mbytes\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "99575"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(df[\"mbytes\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A-427', 'C2BBe1', 'CL-40', 'COLO 741', 'DU145', 'HCC1187', 'HCC1806', 'HCT 116', 'HCT 15', 'HeLa', 'HT115', 'HT-29', 'HT-55', 'Karpas-620', 'KNS-62', 'Ku812', 'LS180', 'MDA-MB-361', 'MDST8', 'NCI-H510A', 'SNU-1', 'SNU-16', 'OVKATE', 'RKO', 'SET-2', 'SiHa', 'Rh30', 'SK-N-SH', 'SNU-C1', 'SR', 'SU-DHL-10', 'SU-DHL-5', 'SU-DHL-6', 'SUIT-2', 'SW1116', 'SW1463', 'SW1573', 'SW403', 'SW48', 'SW620', 'SW780', 'SW837', 'SW948', 'U266B1', 'UACC-812', 'WSU-DLCL2', 'YAPC', 'ZR-75-30']\n"
     ]
    }
   ],
   "source": [
    "r_vector = [\"A-427\", \"C2BBe1\", \"CL-40\", \"COLO 741\", \"DU145\", \"HCC1187\", \"HCC1806\", \"HCT 116\", \"HCT 15\",\n",
    "            \"HeLa\", \"HT115\", \"HT-29\", \"HT-55\", \"Karpas-620\", \"KNS-62\", \"Ku812\", \"LS180\", \"MDA-MB-361\",\n",
    "            \"MDST8\", \"NCI-H510A\", \"SNU-1\", \"SNU-16\", \"OVKATE\", \"RKO\", \"SET-2\", \"SiHa\", \"Rh30\",\n",
    "            \"SK-N-SH\", \"SNU-C1\", \"SR\", \"SU-DHL-10\", \"SU-DHL-5\", \"SU-DHL-6\", \"SUIT-2\", \"SW1116\", \"SW1463\",\n",
    "            \"SW1573\", \"SW403\", \"SW48\", \"SW620\", \"SW780\", \"SW837\", \"SW948\", \"U266B1\", \"UACC-812\",\n",
    "            \"WSU-DLCL2\", \"YAPC\", \"ZR-75-30\"]\n",
    "\n",
    "python_list = [item for item in r_vector]\n",
    "\n",
    "print(python_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>sample_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR8652105</td>\n",
       "      <td>MCF7_BREAST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8639219</td>\n",
       "      <td>DMS53_LUNG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR8652062</td>\n",
       "      <td>IM95_STOMACH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR8652060</td>\n",
       "      <td>KATOIII_STOMACH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR8652054</td>\n",
       "      <td>NCIH2066_LUNG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4545</th>\n",
       "      <td>SRR8657615</td>\n",
       "      <td>NCIH1876_LUNG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4546</th>\n",
       "      <td>SRR8657083</td>\n",
       "      <td>KPNSI9S_AUTONOMIC_GANGLIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4547</th>\n",
       "      <td>SRR8657702</td>\n",
       "      <td>HEC59_ENDOMETRIUM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4548</th>\n",
       "      <td>SRR8657870</td>\n",
       "      <td>TOV21G_OVARY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4549</th>\n",
       "      <td>SRR8657935</td>\n",
       "      <td>SW1463_LARGE_INTESTINE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4550 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             acc                sample_name\n",
       "0     SRR8652105                MCF7_BREAST\n",
       "1     SRR8639219                 DMS53_LUNG\n",
       "2     SRR8652062               IM95_STOMACH\n",
       "3     SRR8652060            KATOIII_STOMACH\n",
       "4     SRR8652054              NCIH2066_LUNG\n",
       "...          ...                        ...\n",
       "4545  SRR8657615              NCIH1876_LUNG\n",
       "4546  SRR8657083  KPNSI9S_AUTONOMIC_GANGLIA\n",
       "4547  SRR8657702          HEC59_ENDOMETRIUM\n",
       "4548  SRR8657870               TOV21G_OVARY\n",
       "4549  SRR8657935     SW1463_LARGE_INTESTINE\n",
       "\n",
       "[4550 rows x 2 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assuming 'df' is your DataFrame\n",
    "subset_df = df[['acc', 'sample_name']]\n",
    "subset_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Match Julia's requested cell lines in the DF to find the SRA accession ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matched entries:\n",
      "Value: A-427\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: C2BBe1\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: CL-40\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: DU145\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: HCC1187\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: HCC1806\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: HeLa\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: HT115\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: HT-29\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: HT-55\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: Karpas-620\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: KNS-62\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: Ku812\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: LS180\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: MDA-MB-361\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: MDST8\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SNU-1\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SNU-16\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: OVKATE\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: RKO\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SET-2\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SiHa\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: Rh30\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SK-N-SH\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SNU-C1\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SR\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SU-DHL-10\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SU-DHL-5\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SU-DHL-6\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SUIT-2\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SW1116\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SW1463\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SW1573\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SW403\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SW48\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SW620\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SW780\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SW837\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: SW948\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: U266B1\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: UACC-812\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: WSU-DLCL2\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: YAPC\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Value: ZR-75-30\n",
      "acc\n",
      "sample_name\n",
      "libraryselection\n",
      "librarysource\n",
      "assay_type\n",
      "\n",
      "Unmatched entries: ['COLO 741', 'HCT 116', 'HCT 15', 'NCI-H510A']\n"
     ]
    }
   ],
   "source": [
    "# Create dictionaries to store matched and unmatched entries\n",
    "matched_entries = {}\n",
    "unmatched_entries = []\n",
    "\n",
    "# Iterate through the r_vector and find matching rows\n",
    "for value in r_vector:\n",
    "    matched_rows = df[df['sample_name'].str.contains(value.replace('-', ''), case=False)]\n",
    "    if not matched_rows.empty:\n",
    "        matched_info = matched_rows[['acc', 'sample_name', 'libraryselection', 'librarysource', 'assay_type']]\n",
    "        matched_entries[value] = matched_rows[['acc', 'sample_name', 'libraryselection', 'librarysource', 'assay_type']].copy()\n",
    "    else:\n",
    "        unmatched_entries.append(value)\n",
    "\n",
    "print(\"Matched entries:\")\n",
    "for value, info in matched_entries.items():\n",
    "    print(f\"Value: {value}\")\n",
    "    for entry in info:\n",
    "        print(entry)\n",
    "    print()\n",
    "\n",
    "print(\"Unmatched entries:\", unmatched_entries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>assay_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>SRR8639203</td>\n",
       "      <td>C2BBE1_LARGE_INTESTINE</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>WGS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>869</th>\n",
       "      <td>SRR8616200</td>\n",
       "      <td>C2BBE1_LARGE_INTESTINE</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2264</th>\n",
       "      <td>SRR8635662</td>\n",
       "      <td>C2BBE1_LARGE_INTESTINE</td>\n",
       "      <td>PCR</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>AMPLICON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3155</th>\n",
       "      <td>SRR8633794</td>\n",
       "      <td>C2BBE1_LARGE_INTESTINE</td>\n",
       "      <td>Reduced Representation</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3685</th>\n",
       "      <td>SRR8657689</td>\n",
       "      <td>C2BBE1_LARGE_INTESTINE</td>\n",
       "      <td>Hybrid Selection</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             acc             sample_name        libraryselection  \\\n",
       "278   SRR8639203  C2BBE1_LARGE_INTESTINE      size fractionation   \n",
       "869   SRR8616200  C2BBE1_LARGE_INTESTINE                    cDNA   \n",
       "2264  SRR8635662  C2BBE1_LARGE_INTESTINE                     PCR   \n",
       "3155  SRR8633794  C2BBE1_LARGE_INTESTINE  Reduced Representation   \n",
       "3685  SRR8657689  C2BBE1_LARGE_INTESTINE        Hybrid Selection   \n",
       "\n",
       "       librarysource        assay_type  \n",
       "278          GENOMIC               WGS  \n",
       "869   TRANSCRIPTOMIC           RNA-Seq  \n",
       "2264         GENOMIC          AMPLICON  \n",
       "3155         GENOMIC     Bisulfite-Seq  \n",
       "3685         GENOMIC  Targeted-Capture  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_entries.get('C2BBe1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'dict' object has no attribute 'to_json'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[56], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m result\u001b[39m=\u001b[39mmatched_entries\u001b[39m.\u001b[39;49mto_json(orient\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mrecords\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m      2\u001b[0m parsed \u001b[39m=\u001b[39m loads(result)\n\u001b[1;32m      3\u001b[0m x\u001b[39m=\u001b[39mdumps(parsed, indent\u001b[39m=\u001b[39m\u001b[39m4\u001b[39m)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'dict' object has no attribute 'to_json'"
     ]
    }
   ],
   "source": [
    "result=matched_entries.to_json(orient=\"records\")\n",
    "parsed = loads(result)\n",
    "x=dumps(parsed, indent=4)\n",
    "\n",
    "filepath=\"./data/JuliaRequest.json\"\n",
    "write_json_to_file(filepath, df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sra_snakemake",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
