{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CCLE RAW DATA EXPLORATION\n",
    "### This notebook is for reading in a set of metadata files from various sources to create a database of metadata + file location for the raw CCLE data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data from the TERRA platform\n",
    "https://app.terra.bio/#workspaces/fccredits-silver-tan-7621/CCLE_v2\n",
    "\n",
    "Each cell contains a link to the google bucket location for each \n",
    "\n",
    "Obtained : August 2023\n",
    "\n",
    "The following raw sequencing data is available for download:\n",
    "\n",
    "- RNAseq (1025) [HG38]\n",
    "- WES (478) [HG38 + HG19]\n",
    "- WGS (329) [HG38 + HG19]\n",
    "- RRBS (unfiltered) (927) [HG19]\n",
    "- HC (976) [HG19]\n",
    "- Raindance (782) [HG19]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entity:sample_id</th>\n",
       "      <th>hg19_RRBS_bai</th>\n",
       "      <th>hg19_RRBS_bam</th>\n",
       "      <th>hg19_hybrid_capture_bai</th>\n",
       "      <th>hg19_hybrid_capture_bam</th>\n",
       "      <th>hg19_raindance_bai</th>\n",
       "      <th>hg19_raindance_bam</th>\n",
       "      <th>hg19_targeted_bai</th>\n",
       "      <th>hg19_targeted_bam</th>\n",
       "      <th>hg38_rna_bai</th>\n",
       "      <th>...</th>\n",
       "      <th>hg38_wes_bai</th>\n",
       "      <th>hg38_wes_bam</th>\n",
       "      <th>hg38_wgs_bai</th>\n",
       "      <th>hg38_wgs_bam</th>\n",
       "      <th>mutect2_parquet_wes</th>\n",
       "      <th>mutect2_parquet_wgs</th>\n",
       "      <th>mutect2_vcf_wes</th>\n",
       "      <th>mutect2_vcf_wgs</th>\n",
       "      <th>participant</th>\n",
       "      <th>stripped_cell_line_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>gs://cclebams/RRBS/G29750/NIH_OVCAR-3/v1/NIH_O...</td>\n",
       "      <td>gs://cclebams/RRBS/G29750/NIH_OVCAR-3/v1/NIH_O...</td>\n",
       "      <td>gs://cclebams/hybrid_capture/NIHOVCAR3_OVARY.bai</td>\n",
       "      <td>gs://cclebams/hybrid_capture/NIHOVCAR3_OVARY.bam</td>\n",
       "      <td>gs://cclebams/raindance/G16640/NIH_OVCAR-3/cur...</td>\n",
       "      <td>gs://cclebams/raindance/G16640/NIH_OVCAR-3/cur...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gs://cclebams/rnasq_hg38/CDS-VqxBGH.Aligned.so...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gs://cclebams/wgs_hg38/CDS-1pLoxn.wgs_ccle.bai</td>\n",
       "      <td>gs://cclebams/wgs_hg38/CDS-1pLoxn.wgs_ccle.bam</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"gs://ccle-mutation/mutect2_parquet/0f5356e34...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gs://ccle-mutation/mutect2_vcf/CDS-1pLoxn_fixe...</td>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>NIHOVCAR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000002</td>\n",
       "      <td>gs://cclebams/RRBS/G29750/HL-60/v2/HL-60.bai</td>\n",
       "      <td>gs://cclebams/RRBS/G29750/HL-60/v2/HL-60.bam</td>\n",
       "      <td>gs://cclebams/hybrid_capture/HL-60.bai</td>\n",
       "      <td>gs://cclebams/hybrid_capture/HL-60.bam</td>\n",
       "      <td>gs://cclebams/raindance/HL-60.bai</td>\n",
       "      <td>gs://cclebams/raindance/HL-60.bam</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gs://cclebams/rnasq_hg38/CDS-AExPJi.Aligned.so...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH-000002</td>\n",
       "      <td>HL60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gs://cclebams/rnasq_hg38/CDS-TyidZO.Aligned.so...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH-000003</td>\n",
       "      <td>CACO2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>gs://cclebams/RRBS/G48942/HEL/current/HEL.bai</td>\n",
       "      <td>gs://cclebams/RRBS/G48942/HEL/current/HEL.bam</td>\n",
       "      <td>gs://cclebams/hybrid_capture/HEL_HAEMATOPOIETI...</td>\n",
       "      <td>gs://cclebams/hybrid_capture/HEL_HAEMATOPOIETI...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gs://cclebams/rnasq_hg38/CDS-Sl0JI6.Aligned.so...</td>\n",
       "      <td>...</td>\n",
       "      <td>gs://cclebams/hg38_wes/CDS-7ffA1d.hg38.bam.bai</td>\n",
       "      <td>gs://cclebams/hg38_wes/CDS-7ffA1d.hg38.bam</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"gs://ccle-mutation/mutect2_parquet/d7afb11fb...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gs://ccle-mutation/mutect2_vcf/CDS-7ffA1d_fixe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>HEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>gs://cclebams/RRBS/G29750/HEL_92.1.7/v2/HEL_92...</td>\n",
       "      <td>gs://cclebams/RRBS/G29750/HEL_92.1.7/v2/HEL_92...</td>\n",
       "      <td>gs://cclebams/hybrid_capture/HEL9217_HAEMATOPO...</td>\n",
       "      <td>gs://cclebams/hybrid_capture/HEL9217_HAEMATOPO...</td>\n",
       "      <td>gs://cclebams/raindance/G16640/HEL_92.1.7/curr...</td>\n",
       "      <td>gs://cclebams/raindance/G16640/HEL_92.1.7/curr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gs://cclebams/rnasq_hg38/CDS-wsWPxR.Aligned.so...</td>\n",
       "      <td>...</td>\n",
       "      <td>gs://cclebams/hg38_wes/CDS-bweujf.hg38.bam.bai</td>\n",
       "      <td>gs://cclebams/hg38_wes/CDS-bweujf.hg38.bam</td>\n",
       "      <td>gs://cclebams/wgs_hg38/CDS-17moNf.wgs_ccle.bai</td>\n",
       "      <td>gs://cclebams/wgs_hg38/CDS-17moNf.wgs_ccle.bam</td>\n",
       "      <td>[\"gs://ccle-mutation/mutect2_parquet/04a2bf1d8...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gs://ccle-mutation/mutect2_vcf/CDS-bweujf_fixe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>HEL9217</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  entity:sample_id                                      hg19_RRBS_bai  \\\n",
       "0       ACH-000001  gs://cclebams/RRBS/G29750/NIH_OVCAR-3/v1/NIH_O...   \n",
       "1       ACH-000002       gs://cclebams/RRBS/G29750/HL-60/v2/HL-60.bai   \n",
       "2       ACH-000003                                                NaN   \n",
       "3       ACH-000004      gs://cclebams/RRBS/G48942/HEL/current/HEL.bai   \n",
       "4       ACH-000005  gs://cclebams/RRBS/G29750/HEL_92.1.7/v2/HEL_92...   \n",
       "\n",
       "                                       hg19_RRBS_bam  \\\n",
       "0  gs://cclebams/RRBS/G29750/NIH_OVCAR-3/v1/NIH_O...   \n",
       "1       gs://cclebams/RRBS/G29750/HL-60/v2/HL-60.bam   \n",
       "2                                                NaN   \n",
       "3      gs://cclebams/RRBS/G48942/HEL/current/HEL.bam   \n",
       "4  gs://cclebams/RRBS/G29750/HEL_92.1.7/v2/HEL_92...   \n",
       "\n",
       "                             hg19_hybrid_capture_bai  \\\n",
       "0   gs://cclebams/hybrid_capture/NIHOVCAR3_OVARY.bai   \n",
       "1             gs://cclebams/hybrid_capture/HL-60.bai   \n",
       "2                                                NaN   \n",
       "3  gs://cclebams/hybrid_capture/HEL_HAEMATOPOIETI...   \n",
       "4  gs://cclebams/hybrid_capture/HEL9217_HAEMATOPO...   \n",
       "\n",
       "                             hg19_hybrid_capture_bam  \\\n",
       "0   gs://cclebams/hybrid_capture/NIHOVCAR3_OVARY.bam   \n",
       "1             gs://cclebams/hybrid_capture/HL-60.bam   \n",
       "2                                                NaN   \n",
       "3  gs://cclebams/hybrid_capture/HEL_HAEMATOPOIETI...   \n",
       "4  gs://cclebams/hybrid_capture/HEL9217_HAEMATOPO...   \n",
       "\n",
       "                                  hg19_raindance_bai  \\\n",
       "0  gs://cclebams/raindance/G16640/NIH_OVCAR-3/cur...   \n",
       "1                  gs://cclebams/raindance/HL-60.bai   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4  gs://cclebams/raindance/G16640/HEL_92.1.7/curr...   \n",
       "\n",
       "                                  hg19_raindance_bam hg19_targeted_bai  \\\n",
       "0  gs://cclebams/raindance/G16640/NIH_OVCAR-3/cur...               NaN   \n",
       "1                  gs://cclebams/raindance/HL-60.bam               NaN   \n",
       "2                                                NaN               NaN   \n",
       "3                                                NaN               NaN   \n",
       "4  gs://cclebams/raindance/G16640/HEL_92.1.7/curr...               NaN   \n",
       "\n",
       "  hg19_targeted_bam                                       hg38_rna_bai  ...  \\\n",
       "0               NaN  gs://cclebams/rnasq_hg38/CDS-VqxBGH.Aligned.so...  ...   \n",
       "1               NaN  gs://cclebams/rnasq_hg38/CDS-AExPJi.Aligned.so...  ...   \n",
       "2               NaN  gs://cclebams/rnasq_hg38/CDS-TyidZO.Aligned.so...  ...   \n",
       "3               NaN  gs://cclebams/rnasq_hg38/CDS-Sl0JI6.Aligned.so...  ...   \n",
       "4               NaN  gs://cclebams/rnasq_hg38/CDS-wsWPxR.Aligned.so...  ...   \n",
       "\n",
       "                                     hg38_wes_bai  \\\n",
       "0                                             NaN   \n",
       "1                                             NaN   \n",
       "2                                             NaN   \n",
       "3  gs://cclebams/hg38_wes/CDS-7ffA1d.hg38.bam.bai   \n",
       "4  gs://cclebams/hg38_wes/CDS-bweujf.hg38.bam.bai   \n",
       "\n",
       "                                 hg38_wes_bam  \\\n",
       "0                                         NaN   \n",
       "1                                         NaN   \n",
       "2                                         NaN   \n",
       "3  gs://cclebams/hg38_wes/CDS-7ffA1d.hg38.bam   \n",
       "4  gs://cclebams/hg38_wes/CDS-bweujf.hg38.bam   \n",
       "\n",
       "                                     hg38_wgs_bai  \\\n",
       "0  gs://cclebams/wgs_hg38/CDS-1pLoxn.wgs_ccle.bai   \n",
       "1                                             NaN   \n",
       "2                                             NaN   \n",
       "3                                             NaN   \n",
       "4  gs://cclebams/wgs_hg38/CDS-17moNf.wgs_ccle.bai   \n",
       "\n",
       "                                     hg38_wgs_bam  \\\n",
       "0  gs://cclebams/wgs_hg38/CDS-1pLoxn.wgs_ccle.bam   \n",
       "1                                             NaN   \n",
       "2                                             NaN   \n",
       "3                                             NaN   \n",
       "4  gs://cclebams/wgs_hg38/CDS-17moNf.wgs_ccle.bam   \n",
       "\n",
       "                                 mutect2_parquet_wes  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3  [\"gs://ccle-mutation/mutect2_parquet/d7afb11fb...   \n",
       "4  [\"gs://ccle-mutation/mutect2_parquet/04a2bf1d8...   \n",
       "\n",
       "                                 mutect2_parquet_wgs  \\\n",
       "0  [\"gs://ccle-mutation/mutect2_parquet/0f5356e34...   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4                                                NaN   \n",
       "\n",
       "                                     mutect2_vcf_wes  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3  gs://ccle-mutation/mutect2_vcf/CDS-7ffA1d_fixe...   \n",
       "4  gs://ccle-mutation/mutect2_vcf/CDS-bweujf_fixe...   \n",
       "\n",
       "                                     mutect2_vcf_wgs participant  \\\n",
       "0  gs://ccle-mutation/mutect2_vcf/CDS-1pLoxn_fixe...  ACH-000001   \n",
       "1                                                NaN  ACH-000002   \n",
       "2                                                NaN  ACH-000003   \n",
       "3                                                NaN  ACH-000004   \n",
       "4                                                NaN  ACH-000005   \n",
       "\n",
       "  stripped_cell_line_name  \n",
       "0               NIHOVCAR3  \n",
       "1                    HL60  \n",
       "2                   CACO2  \n",
       "3                     HEL  \n",
       "4                 HEL9217  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in file at ./terra_combined_paths.csv\n",
    "file_path = './Terra_ExportedList.tsv'\n",
    "terra_df = pd.read_csv(file_path, sep='\\t')\n",
    "terra_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1060 entries, 0 to 1059\n",
      "Data columns (total 21 columns):\n",
      " #   Column                   Non-Null Count  Dtype \n",
      "---  ------                   --------------  ----- \n",
      " 0   entity:sample_id         1060 non-null   object\n",
      " 1   hg19_RRBS_bai            927 non-null    object\n",
      " 2   hg19_RRBS_bam            927 non-null    object\n",
      " 3   hg19_hybrid_capture_bai  973 non-null    object\n",
      " 4   hg19_hybrid_capture_bam  973 non-null    object\n",
      " 5   hg19_raindance_bai       782 non-null    object\n",
      " 6   hg19_raindance_bam       782 non-null    object\n",
      " 7   hg19_targeted_bai        187 non-null    object\n",
      " 8   hg19_targeted_bam        187 non-null    object\n",
      " 9   hg38_rna_bai             1025 non-null   object\n",
      " 10  hg38_rna_bam             1025 non-null   object\n",
      " 11  hg38_wes_bai             474 non-null    object\n",
      " 12  hg38_wes_bam             474 non-null    object\n",
      " 13  hg38_wgs_bai             356 non-null    object\n",
      " 14  hg38_wgs_bam             356 non-null    object\n",
      " 15  mutect2_parquet_wes      474 non-null    object\n",
      " 16  mutect2_parquet_wgs      328 non-null    object\n",
      " 17  mutect2_vcf_wes          474 non-null    object\n",
      " 18  mutect2_vcf_wgs          328 non-null    object\n",
      " 19  participant              1060 non-null   object\n",
      " 20  stripped_cell_line_name  1060 non-null   object\n",
      "dtypes: object(21)\n",
      "memory usage: 174.0+ KB\n"
     ]
    }
   ],
   "source": [
    "terra_df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data from ENA Web Browser\n",
    "https://www.ebi.ac.uk/ena/browser/view/PRJNA523380\n",
    "\n",
    "Obtained: August 2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>secondary_sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>instrument_platform</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>...</th>\n",
       "      <th>study_title</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>fastq_bytes</th>\n",
       "      <th>fastq_md5</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>sra_bytes</th>\n",
       "      <th>sra_md5</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR8615033</td>\n",
       "      <td>SAMN10988141</td>\n",
       "      <td>SRS4395287</td>\n",
       "      <td>SRX5414207</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>Cancer Cell Line Encyclopedia (CCLE)</td>\n",
       "      <td>HCC-56.bam</td>\n",
       "      <td>7773624;10209268</td>\n",
       "      <td>ec71a964d91c08dd4161a5c51e8077ec;f9a75f6d1eb4b...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/003/SRR861...</td>\n",
       "      <td>18104962</td>\n",
       "      <td>be2409df4f8c038b478bdef7d40683bc</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR861/003/SRR8615033</td>\n",
       "      <td>HCC56_LARGE_INTESTINE</td>\n",
       "      <td>HCC56_LARGE_INTESTINE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8615034</td>\n",
       "      <td>SAMN10988287</td>\n",
       "      <td>SRS4395286</td>\n",
       "      <td>SRX5414206</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>Cancer Cell Line Encyclopedia (CCLE)</td>\n",
       "      <td>HCC-78.bam</td>\n",
       "      <td>7530079;9978815</td>\n",
       "      <td>d4fd3f3baaefd30dd1b2ec8157737401;5110277bed499...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/004/SRR861...</td>\n",
       "      <td>17593943</td>\n",
       "      <td>caa4df8f8b40e416f0fa6b7f2c9328ae</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR861/004/SRR8615034</td>\n",
       "      <td>HCC78_LUNG</td>\n",
       "      <td>HCC78_LUNG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR8615035</td>\n",
       "      <td>SAMN10989587</td>\n",
       "      <td>SRS4395285</td>\n",
       "      <td>SRX5414205</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>Cancer Cell Line Encyclopedia (CCLE)</td>\n",
       "      <td>HOP-62.bam</td>\n",
       "      <td>8058229;12501315</td>\n",
       "      <td>0912b268ba684b093da75cec18eea6d0;1112820fb4070...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/005/SRR861...</td>\n",
       "      <td>21029304</td>\n",
       "      <td>7d4124ffdd0d7dc14f3ce2a22ff4c1d2</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR861/005/SRR8615035</td>\n",
       "      <td>HOP62_LUNG</td>\n",
       "      <td>HOP62_LUNG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR8615042</td>\n",
       "      <td>SAMN10987804</td>\n",
       "      <td>SRS4395278</td>\n",
       "      <td>SRX5414198</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>Cancer Cell Line Encyclopedia (CCLE)</td>\n",
       "      <td>NCI-H810.bam</td>\n",
       "      <td>13081808;17856519</td>\n",
       "      <td>89cd15c12b0e3f6d4aa8fa4d930c3768;ec17e81d32c65...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/002/SRR861...</td>\n",
       "      <td>30894571</td>\n",
       "      <td>984e7fe6fbcbdbf30c5e3ae1dd57cd59</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR861/002/SRR8615042</td>\n",
       "      <td>NCIH810_LUNG</td>\n",
       "      <td>NCIH810_LUNG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR8615047</td>\n",
       "      <td>SAMN10987678</td>\n",
       "      <td>SRS4395273</td>\n",
       "      <td>SRX5414193</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>Cancer Cell Line Encyclopedia (CCLE)</td>\n",
       "      <td>NCI-H322.bam</td>\n",
       "      <td>10400525;13057832</td>\n",
       "      <td>40188cfc8cb992e831837517b08b7216;1d9cda042e114...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/007/SRR861...</td>\n",
       "      <td>23711511</td>\n",
       "      <td>648268ec2fe15e87a64f0c0478bf825c</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR861/007/SRR8615047</td>\n",
       "      <td>NCIH322_LUNG</td>\n",
       "      <td>NCIH322_LUNG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession sample_accession secondary_sample_accession  \\\n",
       "0    SRR8615033     SAMN10988141                 SRS4395287   \n",
       "1    SRR8615034     SAMN10988287                 SRS4395286   \n",
       "2    SRR8615035     SAMN10989587                 SRS4395285   \n",
       "3    SRR8615042     SAMN10987804                 SRS4395278   \n",
       "4    SRR8615047     SAMN10987678                 SRS4395273   \n",
       "\n",
       "  experiment_accession study_accession  tax_id scientific_name  \\\n",
       "0           SRX5414207     PRJNA523380    9606    Homo sapiens   \n",
       "1           SRX5414206     PRJNA523380    9606    Homo sapiens   \n",
       "2           SRX5414205     PRJNA523380    9606    Homo sapiens   \n",
       "3           SRX5414198     PRJNA523380    9606    Homo sapiens   \n",
       "4           SRX5414193     PRJNA523380    9606    Homo sapiens   \n",
       "\n",
       "  instrument_platform library_layout library_source  ...  \\\n",
       "0            ILLUMINA         PAIRED        GENOMIC  ...   \n",
       "1            ILLUMINA         PAIRED        GENOMIC  ...   \n",
       "2            ILLUMINA         PAIRED        GENOMIC  ...   \n",
       "3            ILLUMINA         PAIRED        GENOMIC  ...   \n",
       "4            ILLUMINA         PAIRED        GENOMIC  ...   \n",
       "\n",
       "                            study_title     run_alias        fastq_bytes  \\\n",
       "0  Cancer Cell Line Encyclopedia (CCLE)    HCC-56.bam   7773624;10209268   \n",
       "1  Cancer Cell Line Encyclopedia (CCLE)    HCC-78.bam    7530079;9978815   \n",
       "2  Cancer Cell Line Encyclopedia (CCLE)    HOP-62.bam   8058229;12501315   \n",
       "3  Cancer Cell Line Encyclopedia (CCLE)  NCI-H810.bam  13081808;17856519   \n",
       "4  Cancer Cell Line Encyclopedia (CCLE)  NCI-H322.bam  10400525;13057832   \n",
       "\n",
       "                                           fastq_md5  \\\n",
       "0  ec71a964d91c08dd4161a5c51e8077ec;f9a75f6d1eb4b...   \n",
       "1  d4fd3f3baaefd30dd1b2ec8157737401;5110277bed499...   \n",
       "2  0912b268ba684b093da75cec18eea6d0;1112820fb4070...   \n",
       "3  89cd15c12b0e3f6d4aa8fa4d930c3768;ec17e81d32c65...   \n",
       "4  40188cfc8cb992e831837517b08b7216;1d9cda042e114...   \n",
       "\n",
       "                                           fastq_ftp sra_bytes  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/003/SRR861...  18104962   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/004/SRR861...  17593943   \n",
       "2  ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/005/SRR861...  21029304   \n",
       "3  ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/002/SRR861...  30894571   \n",
       "4  ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/007/SRR861...  23711511   \n",
       "\n",
       "                            sra_md5  \\\n",
       "0  be2409df4f8c038b478bdef7d40683bc   \n",
       "1  caa4df8f8b40e416f0fa6b7f2c9328ae   \n",
       "2  7d4124ffdd0d7dc14f3ce2a22ff4c1d2   \n",
       "3  984e7fe6fbcbdbf30c5e3ae1dd57cd59   \n",
       "4  648268ec2fe15e87a64f0c0478bf825c   \n",
       "\n",
       "                                            sra_ftp           sample_alias  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR861/003/SRR8615033  HCC56_LARGE_INTESTINE   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/srr/SRR861/004/SRR8615034             HCC78_LUNG   \n",
       "2  ftp.sra.ebi.ac.uk/vol1/srr/SRR861/005/SRR8615035             HOP62_LUNG   \n",
       "3  ftp.sra.ebi.ac.uk/vol1/srr/SRR861/002/SRR8615042           NCIH810_LUNG   \n",
       "4  ftp.sra.ebi.ac.uk/vol1/srr/SRR861/007/SRR8615047           NCIH322_LUNG   \n",
       "\n",
       "            sample_title  \n",
       "0  HCC56_LARGE_INTESTINE  \n",
       "1             HCC78_LUNG  \n",
       "2             HOP62_LUNG  \n",
       "3           NCIH810_LUNG  \n",
       "4           NCIH322_LUNG  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read ENA_browser_PRJNA523380.tsv file and store in ena_df\n",
    "ena_df = pd.read_csv('./ENA_browser_PRJNA523380.tsv', sep='\\t')\n",
    "ena_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 4550 entries, 0 to 4549\n",
      "Data columns (total 24 columns):\n",
      " #   Column                      Non-Null Count  Dtype \n",
      "---  ------                      --------------  ----- \n",
      " 0   run_accession               4550 non-null   object\n",
      " 1   sample_accession            4550 non-null   object\n",
      " 2   secondary_sample_accession  4550 non-null   object\n",
      " 3   experiment_accession        4550 non-null   object\n",
      " 4   study_accession             4550 non-null   object\n",
      " 5   tax_id                      4550 non-null   int64 \n",
      " 6   scientific_name             4550 non-null   object\n",
      " 7   instrument_platform         4550 non-null   object\n",
      " 8   library_layout              4550 non-null   object\n",
      " 9   library_source              4550 non-null   object\n",
      " 10  read_count                  4550 non-null   int64 \n",
      " 11  base_count                  4550 non-null   int64 \n",
      " 12  first_public                4550 non-null   object\n",
      " 13  last_updated                4550 non-null   object\n",
      " 14  study_title                 4550 non-null   object\n",
      " 15  run_alias                   4550 non-null   object\n",
      " 16  fastq_bytes                 4550 non-null   object\n",
      " 17  fastq_md5                   4550 non-null   object\n",
      " 18  fastq_ftp                   4550 non-null   object\n",
      " 19  sra_bytes                   4550 non-null   int64 \n",
      " 20  sra_md5                     4550 non-null   object\n",
      " 21  sra_ftp                     4550 non-null   object\n",
      " 22  sample_alias                4550 non-null   object\n",
      " 23  sample_title                4550 non-null   object\n",
      "dtypes: int64(4), object(20)\n",
      "memory usage: 853.2+ KB\n"
     ]
    }
   ],
   "source": [
    "# # print out information about dataframes and their columns\n",
    "\n",
    "ena_df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data from SRA RUN SELECTOR\n",
    "https://www.ncbi.nlm.nih.gov/Traces/study/?query_key=2&WebEnv=MCID_64deb5fe5542df760f04c853&o=acc_s%3Aa > Next to 4550 Runs, click \"Metadata\" \n",
    "\n",
    "Obtained: August 2023 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Age</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AssemblyName</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BIOMATERIAL_PROVIDER</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleModel</th>\n",
       "      <th>...</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Sample_type</th>\n",
       "      <th>sex</th>\n",
       "      <th>SRA Study</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>create_date</th>\n",
       "      <th>version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR8633202</td>\n",
       "      <td>66.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>29</td>\n",
       "      <td>1556265367</td>\n",
       "      <td>DSMZ:HUP-T3</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SAMN10987811</td>\n",
       "      <td>Human</td>\n",
       "      <td>...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-03-27T00:00:00Z</td>\n",
       "      <td>HUPT3_PANCREAS</td>\n",
       "      <td>cell culture</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP186687</td>\n",
       "      <td>pancreas</td>\n",
       "      <td>2019-02-27T12:24:00Z</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8633203</td>\n",
       "      <td>74.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>29</td>\n",
       "      <td>1727947890</td>\n",
       "      <td>ATCC:Hs 578T</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SAMN10987893</td>\n",
       "      <td>Human</td>\n",
       "      <td>...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-03-27T00:00:00Z</td>\n",
       "      <td>HS578T_BREAST</td>\n",
       "      <td>cell culture</td>\n",
       "      <td>female</td>\n",
       "      <td>SRP186687</td>\n",
       "      <td>breast</td>\n",
       "      <td>2019-02-27T12:25:00Z</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR8633204</td>\n",
       "      <td>21.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>29</td>\n",
       "      <td>3430374736</td>\n",
       "      <td>ATCC:Hs 604.T</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SAMN10987929</td>\n",
       "      <td>Human</td>\n",
       "      <td>...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-03-27T00:00:00Z</td>\n",
       "      <td>HS604T_FIBROBLAST</td>\n",
       "      <td>cell culture</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP186687</td>\n",
       "      <td>haematopoietic_and_lymphoid_tissue</td>\n",
       "      <td>2019-02-27T12:31:00Z</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR8633205</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>29</td>\n",
       "      <td>1359691158</td>\n",
       "      <td>RIKEN:HuH-6</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SAMN10987713</td>\n",
       "      <td>Human</td>\n",
       "      <td>...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-03-27T00:00:00Z</td>\n",
       "      <td>HUH6_LIVER</td>\n",
       "      <td>cell culture</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP186687</td>\n",
       "      <td>liver</td>\n",
       "      <td>2019-02-27T12:23:00Z</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR8633206</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>29</td>\n",
       "      <td>2005765425</td>\n",
       "      <td>HSSRB</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SAMN10987843</td>\n",
       "      <td>Human</td>\n",
       "      <td>...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-03-27T00:00:00Z</td>\n",
       "      <td>HUH28_BILIARY_TRACT</td>\n",
       "      <td>cell culture</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRP186687</td>\n",
       "      <td>biliary_tract</td>\n",
       "      <td>2019-02-27T12:26:00Z</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Run   Age     Assay Type      AssemblyName  AvgSpotLen       Bases  \\\n",
       "0  SRR8633202  66.0  Bisulfite-Seq  GCA_000001405.13          29  1556265367   \n",
       "1  SRR8633203  74.0  Bisulfite-Seq  GCA_000001405.13          29  1727947890   \n",
       "2  SRR8633204  21.0  Bisulfite-Seq  GCA_000001405.13          29  3430374736   \n",
       "3  SRR8633205   NaN  Bisulfite-Seq  GCA_000001405.13          29  1359691158   \n",
       "4  SRR8633206   NaN  Bisulfite-Seq  GCA_000001405.13          29  2005765425   \n",
       "\n",
       "  BIOMATERIAL_PROVIDER   BioProject     BioSample BioSampleModel  ...  \\\n",
       "0          DSMZ:HUP-T3  PRJNA523380  SAMN10987811          Human  ...   \n",
       "1         ATCC:Hs 578T  PRJNA523380  SAMN10987893          Human  ...   \n",
       "2        ATCC:Hs 604.T  PRJNA523380  SAMN10987929          Human  ...   \n",
       "3          RIKEN:HuH-6  PRJNA523380  SAMN10987713          Human  ...   \n",
       "4                HSSRB  PRJNA523380  SAMN10987843          Human  ...   \n",
       "\n",
       "       Organism  Platform           ReleaseDate          Sample Name  \\\n",
       "0  Homo sapiens  ILLUMINA  2019-03-27T00:00:00Z       HUPT3_PANCREAS   \n",
       "1  Homo sapiens  ILLUMINA  2019-03-27T00:00:00Z        HS578T_BREAST   \n",
       "2  Homo sapiens  ILLUMINA  2019-03-27T00:00:00Z    HS604T_FIBROBLAST   \n",
       "3  Homo sapiens  ILLUMINA  2019-03-27T00:00:00Z           HUH6_LIVER   \n",
       "4  Homo sapiens  ILLUMINA  2019-03-27T00:00:00Z  HUH28_BILIARY_TRACT   \n",
       "\n",
       "    Sample_type     sex  SRA Study                              Tissue  \\\n",
       "0  cell culture    male  SRP186687                            pancreas   \n",
       "1  cell culture  female  SRP186687                              breast   \n",
       "2  cell culture    male  SRP186687  haematopoietic_and_lymphoid_tissue   \n",
       "3  cell culture    male  SRP186687                               liver   \n",
       "4  cell culture     NaN  SRP186687                       biliary_tract   \n",
       "\n",
       "            create_date version  \n",
       "0  2019-02-27T12:24:00Z     1.0  \n",
       "1  2019-02-27T12:25:00Z     1.0  \n",
       "2  2019-02-27T12:31:00Z     1.0  \n",
       "3  2019-02-27T12:23:00Z     1.0  \n",
       "4  2019-02-27T12:26:00Z     1.0  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read ENA_browser_PRJNA523380.tsv file and store in ena_df\n",
    "sra_df = pd.read_csv('./NCBI_SRA_runTable.csv', sep=',')\n",
    "sra_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 4550 entries, 0 to 4549\n",
      "Data columns (total 37 columns):\n",
      " #   Column                Non-Null Count  Dtype  \n",
      "---  ------                --------------  -----  \n",
      " 0   Run                   4550 non-null   object \n",
      " 1   Age                   3573 non-null   float64\n",
      " 2   Assay Type            4550 non-null   object \n",
      " 3   AssemblyName          4550 non-null   object \n",
      " 4   AvgSpotLen            4550 non-null   int64  \n",
      " 5   Bases                 4550 non-null   int64  \n",
      " 6   BIOMATERIAL_PROVIDER  4183 non-null   object \n",
      " 7   BioProject            4550 non-null   object \n",
      " 8   BioSample             4550 non-null   object \n",
      " 9   BioSampleModel        4550 non-null   object \n",
      " 10  Bytes                 4550 non-null   int64  \n",
      " 11  Cell_line             4545 non-null   object \n",
      " 12  Center Name           4550 non-null   object \n",
      " 13  Consent               4550 non-null   object \n",
      " 14  DATASTORE filetype    4550 non-null   object \n",
      " 15  DATASTORE provider    4550 non-null   object \n",
      " 16  DATASTORE region      4550 non-null   object \n",
      " 17  disease               4510 non-null   object \n",
      " 18  disease_stage         4476 non-null   object \n",
      " 19  ETHNICITY             4494 non-null   object \n",
      " 20  Experiment            4550 non-null   object \n",
      " 21  Instrument            4550 non-null   object \n",
      " 22  Isolate               4550 non-null   object \n",
      " 23  Library Name          4550 non-null   object \n",
      " 24  LibraryLayout         4550 non-null   object \n",
      " 25  LibrarySelection      4550 non-null   object \n",
      " 26  LibrarySource         4550 non-null   object \n",
      " 27  Organism              4550 non-null   object \n",
      " 28  Platform              4550 non-null   object \n",
      " 29  ReleaseDate           4550 non-null   object \n",
      " 30  Sample Name           4550 non-null   object \n",
      " 31  Sample_type           4550 non-null   object \n",
      " 32  sex                   4023 non-null   object \n",
      " 33  SRA Study             4550 non-null   object \n",
      " 34  Tissue                4545 non-null   object \n",
      " 35  create_date           4543 non-null   object \n",
      " 36  version               4543 non-null   float64\n",
      "dtypes: float64(2), int64(3), object(32)\n",
      "memory usage: 1.3+ MB\n"
     ]
    }
   ],
   "source": [
    "# print out information about dataframes and their columns\n",
    "sra_df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stripped_cell_line_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NIHOVCAR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HL60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CACO2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HEL9217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055</th>\n",
       "      <td>8305C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1056</th>\n",
       "      <td>8505C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057</th>\n",
       "      <td>PLCPRF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1058</th>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1059</th>\n",
       "      <td>SKNO1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1060 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     stripped_cell_line_name\n",
       "0                  NIHOVCAR3\n",
       "1                       HL60\n",
       "2                      CACO2\n",
       "3                        HEL\n",
       "4                    HEL9217\n",
       "...                      ...\n",
       "1055                   8305C\n",
       "1056                   8505C\n",
       "1057                 PLCPRF5\n",
       "1058                      TT\n",
       "1059                   SKNO1\n",
       "\n",
       "[1060 rows x 1 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "terra_df[['stripped_cell_line_name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_845974/1694021606.py:6: FutureWarning: In a future version of pandas all arguments of StringMethods.split except for the argument 'pat' will be keyword-only.\n",
      "  ena_df[['CELLLINE', 'TISSUE']] = ena_df['sample_alias'].str.split('_', 1, expand=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>secondary_sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>instrument_platform</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>...</th>\n",
       "      <th>fastq_bytes</th>\n",
       "      <th>fastq_md5</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>sra_bytes</th>\n",
       "      <th>sra_md5</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>CELLLINE</th>\n",
       "      <th>TISSUE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR8615033</td>\n",
       "      <td>SAMN10988141</td>\n",
       "      <td>SRS4395287</td>\n",
       "      <td>SRX5414207</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>7773624;10209268</td>\n",
       "      <td>ec71a964d91c08dd4161a5c51e8077ec;f9a75f6d1eb4b...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/003/SRR861...</td>\n",
       "      <td>18104962</td>\n",
       "      <td>be2409df4f8c038b478bdef7d40683bc</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR861/003/SRR8615033</td>\n",
       "      <td>HCC56_LARGE_INTESTINE</td>\n",
       "      <td>HCC56_LARGE_INTESTINE</td>\n",
       "      <td>HCC56</td>\n",
       "      <td>LARGE_INTESTINE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8615034</td>\n",
       "      <td>SAMN10988287</td>\n",
       "      <td>SRS4395286</td>\n",
       "      <td>SRX5414206</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>7530079;9978815</td>\n",
       "      <td>d4fd3f3baaefd30dd1b2ec8157737401;5110277bed499...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/004/SRR861...</td>\n",
       "      <td>17593943</td>\n",
       "      <td>caa4df8f8b40e416f0fa6b7f2c9328ae</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR861/004/SRR8615034</td>\n",
       "      <td>HCC78_LUNG</td>\n",
       "      <td>HCC78_LUNG</td>\n",
       "      <td>HCC78</td>\n",
       "      <td>LUNG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR8615035</td>\n",
       "      <td>SAMN10989587</td>\n",
       "      <td>SRS4395285</td>\n",
       "      <td>SRX5414205</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>8058229;12501315</td>\n",
       "      <td>0912b268ba684b093da75cec18eea6d0;1112820fb4070...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/005/SRR861...</td>\n",
       "      <td>21029304</td>\n",
       "      <td>7d4124ffdd0d7dc14f3ce2a22ff4c1d2</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR861/005/SRR8615035</td>\n",
       "      <td>HOP62_LUNG</td>\n",
       "      <td>HOP62_LUNG</td>\n",
       "      <td>HOP62</td>\n",
       "      <td>LUNG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR8615042</td>\n",
       "      <td>SAMN10987804</td>\n",
       "      <td>SRS4395278</td>\n",
       "      <td>SRX5414198</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>13081808;17856519</td>\n",
       "      <td>89cd15c12b0e3f6d4aa8fa4d930c3768;ec17e81d32c65...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/002/SRR861...</td>\n",
       "      <td>30894571</td>\n",
       "      <td>984e7fe6fbcbdbf30c5e3ae1dd57cd59</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR861/002/SRR8615042</td>\n",
       "      <td>NCIH810_LUNG</td>\n",
       "      <td>NCIH810_LUNG</td>\n",
       "      <td>NCIH810</td>\n",
       "      <td>LUNG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR8615047</td>\n",
       "      <td>SAMN10987678</td>\n",
       "      <td>SRS4395273</td>\n",
       "      <td>SRX5414193</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>10400525;13057832</td>\n",
       "      <td>40188cfc8cb992e831837517b08b7216;1d9cda042e114...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/007/SRR861...</td>\n",
       "      <td>23711511</td>\n",
       "      <td>648268ec2fe15e87a64f0c0478bf825c</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR861/007/SRR8615047</td>\n",
       "      <td>NCIH322_LUNG</td>\n",
       "      <td>NCIH322_LUNG</td>\n",
       "      <td>NCIH322</td>\n",
       "      <td>LUNG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession sample_accession secondary_sample_accession  \\\n",
       "0    SRR8615033     SAMN10988141                 SRS4395287   \n",
       "1    SRR8615034     SAMN10988287                 SRS4395286   \n",
       "2    SRR8615035     SAMN10989587                 SRS4395285   \n",
       "3    SRR8615042     SAMN10987804                 SRS4395278   \n",
       "4    SRR8615047     SAMN10987678                 SRS4395273   \n",
       "\n",
       "  experiment_accession study_accession  tax_id scientific_name  \\\n",
       "0           SRX5414207     PRJNA523380    9606    Homo sapiens   \n",
       "1           SRX5414206     PRJNA523380    9606    Homo sapiens   \n",
       "2           SRX5414205     PRJNA523380    9606    Homo sapiens   \n",
       "3           SRX5414198     PRJNA523380    9606    Homo sapiens   \n",
       "4           SRX5414193     PRJNA523380    9606    Homo sapiens   \n",
       "\n",
       "  instrument_platform library_layout library_source  ...        fastq_bytes  \\\n",
       "0            ILLUMINA         PAIRED        GENOMIC  ...   7773624;10209268   \n",
       "1            ILLUMINA         PAIRED        GENOMIC  ...    7530079;9978815   \n",
       "2            ILLUMINA         PAIRED        GENOMIC  ...   8058229;12501315   \n",
       "3            ILLUMINA         PAIRED        GENOMIC  ...  13081808;17856519   \n",
       "4            ILLUMINA         PAIRED        GENOMIC  ...  10400525;13057832   \n",
       "\n",
       "                                           fastq_md5  \\\n",
       "0  ec71a964d91c08dd4161a5c51e8077ec;f9a75f6d1eb4b...   \n",
       "1  d4fd3f3baaefd30dd1b2ec8157737401;5110277bed499...   \n",
       "2  0912b268ba684b093da75cec18eea6d0;1112820fb4070...   \n",
       "3  89cd15c12b0e3f6d4aa8fa4d930c3768;ec17e81d32c65...   \n",
       "4  40188cfc8cb992e831837517b08b7216;1d9cda042e114...   \n",
       "\n",
       "                                           fastq_ftp sra_bytes  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/003/SRR861...  18104962   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/004/SRR861...  17593943   \n",
       "2  ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/005/SRR861...  21029304   \n",
       "3  ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/002/SRR861...  30894571   \n",
       "4  ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/007/SRR861...  23711511   \n",
       "\n",
       "                            sra_md5  \\\n",
       "0  be2409df4f8c038b478bdef7d40683bc   \n",
       "1  caa4df8f8b40e416f0fa6b7f2c9328ae   \n",
       "2  7d4124ffdd0d7dc14f3ce2a22ff4c1d2   \n",
       "3  984e7fe6fbcbdbf30c5e3ae1dd57cd59   \n",
       "4  648268ec2fe15e87a64f0c0478bf825c   \n",
       "\n",
       "                                            sra_ftp           sample_alias  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR861/003/SRR8615033  HCC56_LARGE_INTESTINE   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/srr/SRR861/004/SRR8615034             HCC78_LUNG   \n",
       "2  ftp.sra.ebi.ac.uk/vol1/srr/SRR861/005/SRR8615035             HOP62_LUNG   \n",
       "3  ftp.sra.ebi.ac.uk/vol1/srr/SRR861/002/SRR8615042           NCIH810_LUNG   \n",
       "4  ftp.sra.ebi.ac.uk/vol1/srr/SRR861/007/SRR8615047           NCIH322_LUNG   \n",
       "\n",
       "            sample_title CELLLINE           TISSUE  \n",
       "0  HCC56_LARGE_INTESTINE    HCC56  LARGE_INTESTINE  \n",
       "1             HCC78_LUNG    HCC78             LUNG  \n",
       "2             HOP62_LUNG    HOP62             LUNG  \n",
       "3           NCIH810_LUNG  NCIH810             LUNG  \n",
       "4           NCIH322_LUNG  NCIH322             LUNG  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Each row in ena_df['sample_alias'] is a concatenation of {CELLLINE}_{TISSUE}\n",
    "# We want to split this into two columns: 'CELLLINE' and 'TISSUE'\n",
    "# We can do this by splitting the string on the first underscore character\n",
    "\n",
    "# Split the sample_alias column on the first underscore character\n",
    "ena_df[['CELLLINE', 'TISSUE']] = ena_df['sample_alias'].str.split('_', 1, expand=True)\n",
    "ena_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_845974/1815170338.py:2: FutureWarning: In a future version of pandas all arguments of StringMethods.split except for the argument 'pat' will be keyword-only.\n",
      "  sra_df[['CELLLINE', 'TISSE']] = sra_df['Sample Name'].str.split('_', 1, expand=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cell_line</th>\n",
       "      <th>CELLLINE</th>\n",
       "      <th>TISSE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HUP-T3</td>\n",
       "      <td>HUPT3</td>\n",
       "      <td>PANCREAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Hs 578T</td>\n",
       "      <td>HS578T</td>\n",
       "      <td>BREAST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hs 604.T</td>\n",
       "      <td>HS604T</td>\n",
       "      <td>FIBROBLAST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HuH-6</td>\n",
       "      <td>HUH6</td>\n",
       "      <td>LIVER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HuH28</td>\n",
       "      <td>HUH28</td>\n",
       "      <td>BILIARY_TRACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4545</th>\n",
       "      <td>DAN-G</td>\n",
       "      <td>DANG</td>\n",
       "      <td>PANCREAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4546</th>\n",
       "      <td>DMS 114</td>\n",
       "      <td>DMS114</td>\n",
       "      <td>LUNG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4547</th>\n",
       "      <td>DMS 273</td>\n",
       "      <td>DMS273</td>\n",
       "      <td>LUNG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4548</th>\n",
       "      <td>QGP-1</td>\n",
       "      <td>QGP1</td>\n",
       "      <td>PANCREAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4549</th>\n",
       "      <td>HCC827</td>\n",
       "      <td>HCC827</td>\n",
       "      <td>LUNG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4550 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Cell_line CELLLINE          TISSE\n",
       "0       HUP-T3    HUPT3       PANCREAS\n",
       "1      Hs 578T   HS578T         BREAST\n",
       "2     Hs 604.T   HS604T     FIBROBLAST\n",
       "3        HuH-6     HUH6          LIVER\n",
       "4        HuH28    HUH28  BILIARY_TRACT\n",
       "...        ...      ...            ...\n",
       "4545     DAN-G     DANG       PANCREAS\n",
       "4546   DMS 114   DMS114           LUNG\n",
       "4547   DMS 273   DMS273           LUNG\n",
       "4548     QGP-1     QGP1       PANCREAS\n",
       "4549    HCC827   HCC827           LUNG\n",
       "\n",
       "[4550 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sra_df already has a cell_line column but its format might not be acceptable so split the Sample Name column\n",
    "sra_df[['CELLLINE', 'TISSE']] = sra_df['Sample Name'].str.split('_', 1, expand=True)\n",
    "sra_df[['Cell_line', 'CELLLINE', 'TISSE']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Age</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AssemblyName</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BIOMATERIAL_PROVIDER</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleModel</th>\n",
       "      <th>...</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Sample_type</th>\n",
       "      <th>sex</th>\n",
       "      <th>SRA Study</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>create_date</th>\n",
       "      <th>version</th>\n",
       "      <th>CELLLINE</th>\n",
       "      <th>TISSE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR8633202</td>\n",
       "      <td>66.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>29</td>\n",
       "      <td>1556265367</td>\n",
       "      <td>DSMZ:HUP-T3</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SAMN10987811</td>\n",
       "      <td>Human</td>\n",
       "      <td>...</td>\n",
       "      <td>2019-03-27T00:00:00Z</td>\n",
       "      <td>HUPT3_PANCREAS</td>\n",
       "      <td>cell culture</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP186687</td>\n",
       "      <td>pancreas</td>\n",
       "      <td>2019-02-27T12:24:00Z</td>\n",
       "      <td>1.0</td>\n",
       "      <td>HUPT3</td>\n",
       "      <td>PANCREAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8633203</td>\n",
       "      <td>74.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>29</td>\n",
       "      <td>1727947890</td>\n",
       "      <td>ATCC:Hs 578T</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SAMN10987893</td>\n",
       "      <td>Human</td>\n",
       "      <td>...</td>\n",
       "      <td>2019-03-27T00:00:00Z</td>\n",
       "      <td>HS578T_BREAST</td>\n",
       "      <td>cell culture</td>\n",
       "      <td>female</td>\n",
       "      <td>SRP186687</td>\n",
       "      <td>breast</td>\n",
       "      <td>2019-02-27T12:25:00Z</td>\n",
       "      <td>1.0</td>\n",
       "      <td>HS578T</td>\n",
       "      <td>BREAST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR8633204</td>\n",
       "      <td>21.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>29</td>\n",
       "      <td>3430374736</td>\n",
       "      <td>ATCC:Hs 604.T</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SAMN10987929</td>\n",
       "      <td>Human</td>\n",
       "      <td>...</td>\n",
       "      <td>2019-03-27T00:00:00Z</td>\n",
       "      <td>HS604T_FIBROBLAST</td>\n",
       "      <td>cell culture</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP186687</td>\n",
       "      <td>haematopoietic_and_lymphoid_tissue</td>\n",
       "      <td>2019-02-27T12:31:00Z</td>\n",
       "      <td>1.0</td>\n",
       "      <td>HS604T</td>\n",
       "      <td>FIBROBLAST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR8633205</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>29</td>\n",
       "      <td>1359691158</td>\n",
       "      <td>RIKEN:HuH-6</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SAMN10987713</td>\n",
       "      <td>Human</td>\n",
       "      <td>...</td>\n",
       "      <td>2019-03-27T00:00:00Z</td>\n",
       "      <td>HUH6_LIVER</td>\n",
       "      <td>cell culture</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP186687</td>\n",
       "      <td>liver</td>\n",
       "      <td>2019-02-27T12:23:00Z</td>\n",
       "      <td>1.0</td>\n",
       "      <td>HUH6</td>\n",
       "      <td>LIVER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR8633206</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>29</td>\n",
       "      <td>2005765425</td>\n",
       "      <td>HSSRB</td>\n",
       "      <td>PRJNA523380</td>\n",
       "      <td>SAMN10987843</td>\n",
       "      <td>Human</td>\n",
       "      <td>...</td>\n",
       "      <td>2019-03-27T00:00:00Z</td>\n",
       "      <td>HUH28_BILIARY_TRACT</td>\n",
       "      <td>cell culture</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRP186687</td>\n",
       "      <td>biliary_tract</td>\n",
       "      <td>2019-02-27T12:26:00Z</td>\n",
       "      <td>1.0</td>\n",
       "      <td>HUH28</td>\n",
       "      <td>BILIARY_TRACT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Run   Age     Assay Type      AssemblyName  AvgSpotLen       Bases  \\\n",
       "0  SRR8633202  66.0  Bisulfite-Seq  GCA_000001405.13          29  1556265367   \n",
       "1  SRR8633203  74.0  Bisulfite-Seq  GCA_000001405.13          29  1727947890   \n",
       "2  SRR8633204  21.0  Bisulfite-Seq  GCA_000001405.13          29  3430374736   \n",
       "3  SRR8633205   NaN  Bisulfite-Seq  GCA_000001405.13          29  1359691158   \n",
       "4  SRR8633206   NaN  Bisulfite-Seq  GCA_000001405.13          29  2005765425   \n",
       "\n",
       "  BIOMATERIAL_PROVIDER   BioProject     BioSample BioSampleModel  ...  \\\n",
       "0          DSMZ:HUP-T3  PRJNA523380  SAMN10987811          Human  ...   \n",
       "1         ATCC:Hs 578T  PRJNA523380  SAMN10987893          Human  ...   \n",
       "2        ATCC:Hs 604.T  PRJNA523380  SAMN10987929          Human  ...   \n",
       "3          RIKEN:HuH-6  PRJNA523380  SAMN10987713          Human  ...   \n",
       "4                HSSRB  PRJNA523380  SAMN10987843          Human  ...   \n",
       "\n",
       "            ReleaseDate          Sample Name   Sample_type     sex  SRA Study  \\\n",
       "0  2019-03-27T00:00:00Z       HUPT3_PANCREAS  cell culture    male  SRP186687   \n",
       "1  2019-03-27T00:00:00Z        HS578T_BREAST  cell culture  female  SRP186687   \n",
       "2  2019-03-27T00:00:00Z    HS604T_FIBROBLAST  cell culture    male  SRP186687   \n",
       "3  2019-03-27T00:00:00Z           HUH6_LIVER  cell culture    male  SRP186687   \n",
       "4  2019-03-27T00:00:00Z  HUH28_BILIARY_TRACT  cell culture     NaN  SRP186687   \n",
       "\n",
       "                               Tissue           create_date version CELLLINE  \\\n",
       "0                            pancreas  2019-02-27T12:24:00Z     1.0    HUPT3   \n",
       "1                              breast  2019-02-27T12:25:00Z     1.0   HS578T   \n",
       "2  haematopoietic_and_lymphoid_tissue  2019-02-27T12:31:00Z     1.0   HS604T   \n",
       "3                               liver  2019-02-27T12:23:00Z     1.0     HUH6   \n",
       "4                       biliary_tract  2019-02-27T12:26:00Z     1.0    HUH28   \n",
       "\n",
       "           TISSE  \n",
       "0       PANCREAS  \n",
       "1         BREAST  \n",
       "2     FIBROBLAST  \n",
       "3          LIVER  \n",
       "4  BILIARY_TRACT  \n",
       "\n",
       "[5 rows x 39 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sra_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "run_accession                                                        SRR8615081\n",
       "sample_accession                                                   SAMN10987705\n",
       "secondary_sample_accession                                           SRS4395239\n",
       "experiment_accession                                                 SRX5414159\n",
       "study_accession                                                     PRJNA523380\n",
       "tax_id                                                                     9606\n",
       "scientific_name                                                    Homo sapiens\n",
       "instrument_platform                                                    ILLUMINA\n",
       "library_layout                                                           PAIRED\n",
       "library_source                                                          GENOMIC\n",
       "read_count                                                                10818\n",
       "base_count                                                              3245400\n",
       "first_public                                                         2019-03-28\n",
       "last_updated                                                         2019-03-28\n",
       "study_title                                Cancer Cell Line Encyclopedia (CCLE)\n",
       "run_alias                                                        ACC-MESO-1.bam\n",
       "fastq_bytes                                                       525302;830170\n",
       "fastq_md5                     21ea8c10bc7ca9362afe39adbae80d07;002ddc347a454...\n",
       "fastq_ftp                     ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/001/SRR861...\n",
       "sra_bytes                                                               1571192\n",
       "sra_md5                                        dc571eaab32ec6119c2fb7c7cc1525f2\n",
       "sra_ftp                        ftp.sra.ebi.ac.uk/vol1/srr/SRR861/001/SRR8615081\n",
       "sample_alias                                                    ACCMESO1_PLEURA\n",
       "sample_title                                                    ACCMESO1_PLEURA\n",
       "CELLLINE                                                               ACCMESO1\n",
       "TISSUE                                                                   PLEURA\n",
       "Name: 400, dtype: object"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the row with the lowest sra_bytes in ena_df\n",
    "ena_df.loc[ena_df['sra_bytes'].idxmin()]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "314    8680933841\n",
       "Name: sra_bytes, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get row where run_accession is SRR8615788\n",
    "ena_df.loc[ena_df['run_accession'] == 'SRR8615788']['sra_bytes']\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "snakemake",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
